### ISSUE TITLE
Download of channel response information

### ISSUE THREAD

**Issue by mmoorkamp** [fmagrini/seislib#6]
Hi,

I have recently discovered seislib and it seems an excellent tool for what I am planning to do. Unfortunately, I am running into an issue when preparing to download data. I am trying to get a reasonably large set of seismic stations from the western United States (although the same error sometimes occurs when I restrict the number of stations).

Based on the wiki, I am running the following code

 stations_config=dict(
        channel='LH*,BH*,HH*',
        includerestricted=False,        
        maxlatitude=49,
        minlatitude=38,
        minlongitude=-125,
        maxlongitude=-105)   

from obspy import UTCDateTime as UTC
from obspy.geodetics import degrees2kilometers

events_config=dict(
        starttime=UTC(2010, 1, 1),
        endtime=UTC(2020, 1, 1),
        depthmax=50,
        magmin=6, 
        magmax=8.5,
        distmin=degrees2kilometers(20),
        distmax=degrees2kilometers(140),
            ) 

from seislib.eq import EQDownloader
src = '/home/moorkamp/SWData/data'
downloader = EQDownloader(savedir=src,
                          inv_provider='iris',
                          ev_provider='iris',
                          inventory_name='iris.xml',
                          sampling_rate=1,
                          prefilter=(0.001, 0.005, 0.1, 0.4),
                          vmin=1.5,
                          vmax=5.5,
                          units='disp',
                          attach_response=False,
                          stations_config=stations_config,
                          events_config=events_config,
                          verbose=True)        
This last command, gives me errors. I have been able to trace it to the following lines in eq_downloader.py (345 - 369)

        def attach_channels(inv, **kwargs):
            tmp_kwargs = {i:j for i, j in kwargs.items() \
                          if i not in ['network', 'station']}
            for network in inv:
                for station in network:
                    if station.__dict__.get('creation_date') is None:
                        station.creation_date = station.start_date
                    while 1:
                        try:
                            print ('Getting station', station.code )
                            tmp = self.staclient.get_stations(network=network.code, 
                                                              station=station.code, 
                                                              level='response',
                                                              **tmp_kwargs)
                            station.channels.extend(tmp[0][0].channels)

                            break
                        except FDSNNoDataException:
                            break
                        except FDSNException:
                            time.sleep(1)
                            continue
            return inv

Here, the line 

                            tmp = self.staclient.get_stations(network=network.code, 
                                                              station=station.code, 
                                                              level='response',
                                                              **tmp_kwargs)

fails at random stations. Lookign further into the obspy code, it seems that the information request is acknowledged by the IRIS FDSN web service and returns that everything is OK (http code 200). However, the following parsing of the information fails, line 738ff in client.py from obspy

        if "station" not in self.services:
            msg = "The current client does not have a station service."
            raise ValueError(msg)

        locs = locals()
        setup_query_dict('station', locs, kwargs)

        url = self._create_url_from_parameters(
            "station", DEFAULT_PARAMETERS['station'], kwargs)
        self.debug = True
        data_stream = self._download(url,use_gzip=False)
        data_stream.seek(0, 0)

        if filename:
            self._write_to_file_object(filename, data_stream)
            data_stream.close()
        else:
            # This works with XML and StationXML data.
            if format is None or format == 'xml':
                inventory = read_inventory(data_stream, format='STATIONXML')
            elif format == 'text':
                inventory = read_inventory(data_stream, format='STATIONTXT')
            else:
                inventory = read_inventory(data_stream)
            data_stream.close()
            return inventory

Is this a known problem? Does IRIS have a stability issue when trying to get information about many stations? What is the best way to handle this?

Any help is appreciated.

Regards

Max

**Comment by fmagrini** [comment]
Hi Max,

Thanks for opening the issue. I have just tried to run your code, and I indeed get the following error

```
Traceback (most recent call last):

  File "<ipython-input-6-a4993517e695>", line 13, in <module>
    verbose=True)

  File "/home/fabrizio/anaconda3/lib/python3.7/site-packages/seislib/eq/eq_downloader.py", line 304, in __init__
    self.inventory = self.build_inventory(**stations_config)

  File "/home/fabrizio/anaconda3/lib/python3.7/site-packages/seislib/eq/eq_downloader.py", line 372, in build_inventory
    return attach_channels(inv, **kwargs)

  File "/home/fabrizio/anaconda3/lib/python3.7/site-packages/seislib/eq/eq_downloader.py", line 357, in attach_channels
    **tmp_kwargs)

  File "/home/fabrizio/anaconda3/lib/python3.7/site-packages/obspy/clients/fdsn/client.py", line 718, in get_stations
    inventory = read_inventory(data_stream)

  File "<decorator-gen-171>", line 2, in read_inventory

  File "/home/fabrizio/anaconda3/lib/python3.7/site-packages/obspy/core/util/decorator.py", line 301, in _map_example_filename
    return func(*args, **kwargs)

  File "/home/fabrizio/anaconda3/lib/python3.7/site-packages/obspy/core/inventory/inventory.py", line 98, in read_inventory
    format=format, *args, **kwargs)[0]

  File "/home/fabrizio/anaconda3/lib/python3.7/site-packages/obspy/core/util/base.py", line 386, in _read_from_plugin
    is_format = is_format(filename)

  File "/home/fabrizio/anaconda3/lib/python3.7/site-packages/obspy/io/stationtxt/core.py", line 110, in is_fdsn_station_text_file
    if not first_line.startswith("#"):

TypeError: startswith first arg must be bytes or a tuple of bytes, not str
```

This is the first time I run into something similar. May you please confirm this is exactly the same error you get?

I will have a look and get back to you.
Cheers

**Comment by mmoorkamp** [comment]
Hi Fabrizio,
mine looks slightly differently

Traceback (most recent call last):
  File "/home/moorkamp/SWData/./seislib_tele.py", line 61, in <module>
    downloader = EQDownloader(savedir=src,
  File "/home/moorkamp/.local/lib/python3.10/site-packages/seislib/eq/eq_downloader.py", line 304, in __init__
    self.inventory = self.build_inventory(**stations_config)
  File "/home/moorkamp/.local/lib/python3.10/site-packages/seislib/eq/eq_downloader.py", line 376, in build_inventory
    return attach_channels(inv, **kwargs)
  File "/home/moorkamp/.local/lib/python3.10/site-packages/seislib/eq/eq_downloader.py", line 356, in attach_channels
    tmp = self.staclient.get_stations(network=network.code, 
  File "/home/moorkamp/.local/lib/python3.10/site-packages/obspy/clients/fdsn/client.py", line 755, in get_stations
    inventory = read_inventory(data_stream, format='STATIONXML')
  File "<decorator-gen-49>", line 2, in read_inventory
  File "/home/moorkamp/.local/lib/python3.10/site-packages/obspy/core/util/decorator.py", line 297, in _map_example_filename
    return func(*args, **kwargs)
  File "/home/moorkamp/.local/lib/python3.10/site-packages/obspy/core/inventory/inventory.py", line 92, in read_inventory
    return _generic_reader(path_or_file_object, _read, format=format,
  File "/home/moorkamp/.local/lib/python3.10/site-packages/obspy/core/util/base.py", line 624, in _generic_reader
    generic = callback_func(pathname_or_url, **kwargs)
  File "<decorator-gen-50>", line 2, in _read
  File "/home/moorkamp/.local/lib/python3.10/site-packages/obspy/core/util/decorator.py", line 142, in uncompress_file
    return func(filename, *args, **kwargs)
  File "/home/moorkamp/.local/lib/python3.10/site-packages/obspy/core/inventory/inventory.py", line 101, in _read
    inventory, format = _read_from_plugin('inventory', filename, format=format,
  File "/home/moorkamp/.local/lib/python3.10/site-packages/obspy/core/util/base.py", line 423, in _read_from_plugin
    list_obj = read_format(filename, **kwargs)
  File "/home/moorkamp/.local/lib/python3.10/site-packages/obspy/io/stationxml/core.py", line 152, in _read_stationxml
    root = etree.parse(path_or_file_object).getroot()
  File "src/lxml/etree.pyx", line 3536, in lxml.etree.parse
  File "src/lxml/parser.pxi", line 1893, in lxml.etree._parseDocument
  File "src/lxml/parser.pxi", line 1913, in lxml.etree._parseMemoryDocument
  File "src/lxml/parser.pxi", line 1800, in lxml.etree._parseDoc
  File "src/lxml/parser.pxi", line 1141, in lxml.etree._BaseParser._parseDoc
  File "src/lxml/parser.pxi", line 615, in lxml.etree._ParserContext._handleParseResultDoc
  File "src/lxml/parser.pxi", line 725, in lxml.etree._handleParseResult
  File "src/lxml/parser.pxi", line 654, in lxml.etree._raiseParseError
  File "<string>", line 1
lxml.etree.XMLSyntaxError: Start tag expected, '<' not found, line 1, column 1

The minor differences in line numbers eq_downloader.py probably come from some print statements that I added to the code for debugging purposes. You also seem to have a slightly different version of obspy. I have 1.4.0.

Hope this helps

Max

**Comment by fmagrini** [comment]
I released a new patch (0.6.3) that fixes the error I copy-pasted above. It was indeed a very weird error, meaning that my knowledge of the relevant obspy module is not deep enough to understand why it did not occur systematically.

Unfortunately, I was not able to produce your error, instead. I tried to do so through both python 3.7 + obspy 1.1.0 and python 3.9 + obspy 1.2.2.

I suggest that you try to install the latest version of SeisLib. If you still have the same error, I suggest that you create a new environment (assuming you use conda) following the SeisLib documentation (https://seislib.readthedocs.io/en/latest/installation.html) so as to run seislib along with Python 3.9. 

Please let me know if you have additional problems and eventually close this issue.

Thank you very much,
Fabrizio
